### PR TITLE
Scheduled DB Query for Email Cron Job

### PR DESCRIPTION
**Pull request by victoranyun**
### Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/Cron-job-to-send-reminder-emails-4f2a2a05adba489bbf75d66d7c69a362)

### Implementation description

- Set up AWS Lambda to ping backend every hour @ /api/mail
- Endpoint checks DB for any classes that are starting in 2-4 hours from the current time and 47-49 hours from the current time
- Sends email to all volunteers/parents registered for these classes

Screen Shot 2021-08-19 at 2 36 00 AM
Screen Shot 2021-08-19 at 2 36 46 AM
Screen Shot 2021-08-19 at 2 35 38 AM

Screen Shot 2021-08-29 at 6 31 56 PM

### Steps to test

1. Setup records in local DB
2. /api/mail

### What should reviewers focus on?

- Email template and what contents and how to abstract

### Checklist

-   [x] My PR name is descriptive and in imperative tense
-   [x] I have run the linter
-   [x] I have requested a review from the PL, and/or other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts]: Make 3 a constant

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts]: Make 48 a constant so it's more clear what this number means

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts]: Thoughts on using Promise.all for handling bulk email sending?

https://github.com/uwblueprint/dancefest-web/blob/6dbe38f8d950e4cf5c05911acb04752596d6838a/pages/api/feedback/index.js#L94-L96

File: services/database/mail.ts
Comment (feng-eric) [inline services/database/mail.ts]:
```suggestion
    const oneHourAfter = new Date();
```

File: services/database/mail.ts
Comment (feng-eric) [inline services/database/mail.ts]:
```suggestion
    const  oneHourBefore = new Date();
```

File: services/database/mail.ts
Comment (feng-eric) [inline services/database/mail.ts:35]: Add a comment to make it clear the range we're querying for

File: utils/sendMail.ts
Comment (feng-eric) [inline utils/sendMail.ts]: We have `services/nodemailer.mail.ts`. Should we get rid of that if we're going to use this?

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts]: Any particular reason we're using `for` loops instead of `foreach`?

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts]: Let's add comments saying what all these loops are doing. For example, this is sending all volunteer registrants a 3 hour email reminder.

File: pages/api/mail/index.ts
Comment (feng-eric) [reply pages/api/mail/index.ts]: Also, `result3h` isn't really descriptive. What you're returning is a list of classes joined together w other stuff right?

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts]: How come there isn't a subject and content in these emails?

Comment (feng-eric) [reply pages/api/mail/index.ts]: If it's pretty much the same thing as the one above, perhaps you can make it a constant or something and just have the name and hours be different.

File: services/database/mail.ts
Comment (feng-eric) [inline services/database/mail.ts:8]: Please add comments to describe what this function is doing, what it takes in, and what it returns.

**Comment by feng-eric** [comment]
Could you also include a screenshot of the AWS Lambda function code so reviewers can see it?

File: pages/api/mail/index.ts
Comment (victoranyun) [reply pages/api/mail/index.ts]: I'm not sure how to use foreach with the indexing of the parentRegs and volunteerRegs

File: utils/sendMail.ts
Comment (victoranyun) [reply utils/sendMail.ts]: Done

File: pages/api/mail/index.ts
Comment (victoranyun) [reply pages/api/mail/index.ts]: Done

**Comment by victoranyun** [comment]
> Could you also include a screenshot of the AWS Lambda function code so reviewers can see it?

Done

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts]:
```suggestion
```

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts]: Since this function is being called every hour, is it possible that duplicate emails get sent using this time range?

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts]: Perhaps we can make `classesInThreeHours[i].parentRegs` a constant so it's more readable? For some of these nested loops, the names are pretty long.

File: services/database/mail.ts
Comment (cindy-wang328) [inline services/database/mail.ts]: I'm kinda confused why it's hoursWIthin +/- 1 haha
Also should one of these be not inclusive? Intuitively, i think if both ends of the range are inclusive you might get duplicates when something equals a range boundary

Also i kinda agree with Eric's comment on how to make sure you don't get duplicates?
Like if something has startDate 7, and hoursWithin is 1
At time 5, oneHourBefore is 5 + (hoursWithin - 1) = 5, oneHourAfter is 5 + (hoursWithin + 1) = 7, so if there were startDates with 7 then it'll be included
(since it runs every hour) at time 7, oneHourBefore is 7 and oneHourAfter is 9, then the startDate of 7 will get included again?

Comment (cindy-wang328) [reply services/database/mail.ts]: Clarification:
Right now, hoursWithin +/- 1 means the range is 2h
I think if it runs every 1h, the range should be 1h (so theres no overlap)

I messaged Eric and we thought the bounds should be, >= hoursWithin and < hoursWithin + 1

like if rn the time is 0
and you have A=3, B=3.5, C=4
and hoursWithin is 3
Then for t=0, the range is >= 3 and < 4, you send A and B
for t=1, the range is >= 4 and < 5, you send C

File: services/database/mail.ts
Comment (cindy-wang328) [inline services/database/mail.ts:11]: Why are we doing one hour after and before :O I'm curious

File: pages/api/mail/index.ts
Comment (mahadzaryab1) [inline pages/api/mail/index.ts]: nit: Variable names should be related to what those values contain and not the value itself (ex. `firstIntervalHours`). If we decide to change the value it shouldn't impact the variable name.

File: pages/api/mail/index.ts
Comment (mahadzaryab1) [inline pages/api/mail/index.ts]: nit: construct this as a variable rather than passing it in as a big string literal

File: pages/api/mail/index.ts
Comment (mahadzaryab1) [inline pages/api/mail/index.ts]: same comment as  above

File: pages/api/mail/index.ts
Comment (mahadzaryab1) [inline pages/api/mail/index.ts]: this exact block seems to be repeated a few times - can we make a helper function with parameters passed in?

**Review by mahadzaryab1 — COMMENTED** [review]
added a few comments

File: pages/api/mail/index.ts
Comment (feng-eric) [reply pages/api/mail/index.ts]: Is your comment here still valid or is it outdated?

File: services/database/mail.ts
Comment (feng-eric) [reply services/database/mail.ts]: @victoranyun Should it be `lte` or `le`?

**Review by feng-eric — COMMENTED** [review]
+1 to @mahadzaryab1's suggestion of using a helper function for the repeated code blocks. 

Also, can you resolve the comments that you have fixed? It'll be easier to know what you've addressed so far @victoranyun

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts]: nit: Let's add helpers at the bottom of files

File: services/database/mail.ts
Comment (feng-eric) [inline services/database/mail.ts]: Update the comment.

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts]: I don't think it makes sense to hard code 3 hours since we're using this helper function for 48 hours too. I think you should pass in the const variable for number of hours and use it here: `${...}`

Comment (feng-eric) [reply pages/api/mail/index.ts]: Let's change it for the above one too.

**Review by feng-eric — COMMENTED** [review]
Were you able to test these new changes btw?

**Comment by victoranyun** [comment]
> Were you able to test these new changes btw?

Yes, I updated the changes like you mentioned and I tested it

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts:20]: Perhaps have a comment explaining the need of the AWS_LAMBDA_KEY env variable.

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts]: Should we include the specific class name in the email subject?

Comment (victoranyun) [reply pages/api/mail/index.ts]: Not too sure, I feel like we should include "Social Diversity for Children" though but if we include the class name it might be too long of a subject? I don't know though

**Review by feng-eric — APPROVED** [review]
LGTM as long as you have tested it thoroughly. Perhaps test it once more just to re-confirm. Let's hold off on merging until after the evening today (client meeting is 6-6:45 ET) @victoranyun

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts]: Should these two functions be in `utils`? They seem to be generic enough to be potentially used somewhere else too. Thoughts? @victoranyun

File: pages/api/mail/index.ts
Comment (feng-eric) [inline pages/api/mail/index.ts:84]: Can you add a screenshot of this email in the PR? Thanks!

**Pull request merged**
